In [ ]:
import kagglehub

# Download latest version
path = kagglehub.dataset_download("erqizhou/students-data-analysis")

print("Path to dataset files:", path)

100%|██████████| 2.12k/2.12k [00:00<00:00, 2.49MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/erqizhou/students-data-analysis/versions/1


In [ ]:
import os

dataset_path = '/root/.cache/kagglehub/datasets/erqizhou/students-data-analysis/versions/1'
files = os.listdir(dataset_path)

print("Daftar File dalam Folder:")
for f in files:
    print("-", f)

Daftar File dalam Folder:
- Students data.csv


In [ ]:
import pandas as pd
data_files = os.path.join(dataset_path, 'Students data.csv')
data = pd.read_csv(data_files)
data

ID class  gender  race    GPA  Algebra  Calculus1  Calculus2  \
0    1141     A    male     1  73.47       64         81         87   
1    1142     A  female     1  71.22       57         50         51   
2    1143     A  female     2  74.56       47         48         71   
3    1144     A  female     1  72.89       46         72         38   
4    1145     A  female     1  70.11       49         45         63   
..    ...   ...     ...   ...    ...      ...        ...        ...   
100  1241     A  female     1  88.34       87         83         92   
101  1242     B    male     1  89.84       98         77         95   
102  1243     B    male     1  88.82       83         80         91   
103  1244     A    male     1  86.60       92         82         91   
104  1245     A    male     1  93.71       93         97         99   

     Statistics  Probability  Measure  Functional_analysis from1 from2 from3  \
0            60           74       71                   60     A     A     A   
1            51           55       62                   61     B     A     A   
2            60           61       68                   64     C     A     A   
3            60           29       54                   51     D     A     A   
4            60           66       66                   61     E     A     A   
..          ...          ...      ...                  ...   ...   ...   ...   
100          98           93       86                   90     M     B     A   
101          98           96       88                  100     A     B     A   
102          98           93       95                   71     T     B     A   
103          99           94       82                   78     S     B     A   
104         100           97       90                   90     K     B     A   

     from4  y  
0        3  0  
1        2  0  
2        0  1  
3        0  0  
4        0  0  
..     ... ..  
100      0  1  
101      0  1  
102      0  2  
103      0  2  
104      0  2  

[105 rows x 17 columns]

PCA merupakan metode statistik yang digunakan untuk <b>Mereduksi Dimensi</b> dengan cara mengubah data ke dalam komponen utama (Principal Component) yang memiliki variansi terbesar. PCA digunakan untuk mengatasi masalah dimensi pada data yang memiliki banyak fitur, tanpa kehilangan terlalu banyak informasi penting.

Manfaat PCA:

1. Reduksi Dimensi

2. Visualisasi Data

3. Mengurangi Overfitting (menghapus fitur yang kurang penting

4. Mempercepat Komputasi

##Perbedaan akurasi model dengan menggunakan 3 data dan 5 data yang direduksi menjadi 3 data menggunakan PCA

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

mengambil 3 data secara langsung (tanpa pca)

In [ ]:
x = data[['Algebra', 'Calculus1']]
y = data[['y']]

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=3
)

In [ ]:
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

model = SVC(kernel='sigmoid')
model.fit(X_train, y_train.values.ravel())

y_pred = model.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.3333333333333333


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score
import matplotlib.pyplot as plt

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=3)


model = SVC(kernel='poly')
model.fit(x_train, y_train.values.ravel())

y_pred_train = model.predict(x_train)
y_pred_test = model.predict(x_test)

acc_train = accuracy_score(y_train, y_pred_train)
print("Accuracy Training =", acc_train)

acc_test = accuracy_score(y_test, y_pred_test)
print("Accuracy Testing =", acc_test)

Accuracy Training = 0.6309523809523809
Accuracy Testing = 0.47619047619047616


###mengambil 5 data dan kemudian menggunakan pca menjadi 3 data

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 105 entries, 0 to 104
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   105 non-null    int64  
 1   class                105 non-null    object 
 2   gender               105 non-null    object 
 3   race                 105 non-null    int64  
 4   GPA                  105 non-null    float64
 5   Algebra              105 non-null    int64  
 6   Calculus1            105 non-null    int64  
 7   Calculus2            105 non-null    int64  
 8   Statistics           105 non-null    int64  
 9   Probability          105 non-null    int64  
 10  Measure              105 non-null    int64  
 11  Functional_analysis  105 non-null    int64  
 12  from1                105 non-null    object 
 13  from2                105 non-null    object 
 14  from3                105 non-null    object 
 15  from4                105 non-null    int

In [ ]:
x = data[['Algebra', 'Calculus1','Calculus2','Statistics']]
y = data[['y']]

In [ ]:

from sklearn.decomposition import PCA

from sklearn.preprocessing import StandardScaler

In [ ]:

#Pre proceesing
#Normalisasi Data (Data di-scaling agar memiliki varians yang seragan)

scaler = StandardScaler()

x_scaled =scaler.fit_transform(x)
x_scaled

array([[-1.03346973e+00,  7.44561324e-01, -2.45911240e+00],
       [-1.63346993e+00, -1.80922910e+00, -3.33969642e+00],
       [-2.49061307e+00, -1.97398977e+00, -2.45911240e+00],
       [-2.57632739e+00,  3.13829852e-03, -2.45911240e+00],
       [-2.31918445e+00, -2.22113078e+00, -2.45911240e+00],
       [-1.37632699e+00, -9.85425735e-01, -2.36126973e+00],
       [-1.37632699e+00, -2.38589145e+00, -2.16558440e+00],
       [-1.37632699e+00, -2.79779313e+00, -2.16558440e+00],
       [-1.29061267e+00, -9.85425735e-01, -1.67637105e+00],
       [-1.37632699e+00, -9.03045399e-01, -1.57852839e+00],
       [-1.29061267e+00, -4.91143718e-01, -1.57852839e+00],
       [-2.62040904e-01,  2.50279307e-01, -1.38284305e+00],
       [-5.19183846e-01, -9.85425735e-01, -1.38284305e+00],
       [-7.76326789e-01, -5.73524055e-01, -1.18715771e+00],
       [-2.62040904e-01, -9.85425735e-01, -8.93629706e-01],
       [-3.47755218e-01, -7.38284727e-01, -8.93629706e-01],
       [-1.76326589e-01, -1.61622374e-01

In [ ]:
pca = PCA(n_components = 2) #mengurangi 4 fitur menjadi 2 fitur

x_pca = pca.fit_transform(x_scaled)

In [ ]:
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score

x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.2, random_state=3)

model = SVC()
model.fit(x_train, y_train.values.ravel())

y_pred_train = model.predict(x_train)
y_pred_test = model.predict(x_test)

print("Accuracy Training =", accuracy_score(y_train, y_pred_train))
print("Accuracy Testing =", accuracy_score(y_test, y_pred_test))

Accuracy Training = 0.6190476190476191
Accuracy Testing = 0.38095238095238093


In [ ]:
x = data[['Algebra', 'Calculus1', 'Calculus2', 'Statistics', 'Probability', 'Measure', 'Functional_analysis']]
y = data[['y']]

In [ ]:
scaler = StandardScaler()

x_scaled =scaler.fit_transform(x)
x_scaled

array([[-1.03346973e+00,  7.44561324e-01,  5.39815336e-01,
        -2.45911240e+00, -9.43809765e-01, -9.52662300e-01,
        -1.18410530e+00],
       [-1.63346993e+00, -1.80922910e+00, -1.87212553e+00,
        -3.33969642e+00, -2.75952865e+00, -1.83097047e+00,
        -1.10683305e+00],
       [-2.49061307e+00, -1.97398977e+00, -5.32158381e-01,
        -2.45911240e+00, -2.18614374e+00, -1.24543169e+00,
        -8.75016286e-01],
       [-2.57632739e+00,  3.13829852e-03, -2.74310417e+00,
        -2.45911240e+00, -5.24419659e+00, -2.61168884e+00,
        -1.87955559e+00],
       [-2.31918445e+00, -2.22113078e+00, -1.06814524e+00,
        -2.45911240e+00, -1.70832298e+00, -1.44061128e+00,
        -1.10683305e+00],
       [-1.37632699e+00, -9.85425735e-01, -2.67610581e+00,
        -2.36126973e+00, -1.80388713e+00, -6.59892910e-01,
        -1.18410530e+00],
       [-1.37632699e+00, -2.38589145e+00, -1.80512717e+00,
        -2.16558440e+00, -1.23050222e+00, -8.55072503e-01,
        -2.5021491

In [ ]:
pca = PCA(n_components = 4) #mengurangi 4 fitur menjadi 2 fitur

x_pca = pca.fit_transform(x_scaled)

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x_pca, y, test_size=0.2, random_state=3)

In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.svm import SVC

model_svc = SVC()
model_svc.fit(x_train, y_train)

y_pred_train = model_svc.predict(x_train)
y_pred_test = model_svc.predict(x_test)

akurasi_training = accuracy_score(y_train, y_pred_train)
akurasi_testing = accuracy_score(y_test, y_pred_test)
print(f'Hasil Akurasi Training = {akurasi_training:.2%}')
print(f'Hasil Akurasi Testing = {akurasi_testing:.2%}')

Hasil Akurasi Training = 63.10%
Hasil Akurasi Testing = 42.86%


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


###Berdasarkan hasil Pertama, Data yang digunakan untuk traiing hanya 2 fitur saja memberikan akurasi sebesar 61.90% atau (0.6190. Sedangkan data dengan 7 fitur kemudian direduksi menjadi 2 fitur agar informasi data sebelumnya tidak hilang, memberikan hasil akurasi sebesar 63.10%.

###Kesimpulannya:<br>
Untuk dapat mempertahankan tanpa mengurangi akurasi yang besar, disarankan untuk mengambil fitur sebanyak-banyaknya dan melakukan reduksi dimensi menggunakan PCA, juga jika tujuannya adalah visualisasi 2 dimensi (2 fitur) atau 3 dimensi (3 fitur)